x/wasm/keeper/keeper.go

```diff
@@ -565,7 +567,30 @@ func (k Keeper) callMigrateEntrypoint(
 	prefixStoreKey := types.GetContractStorePrefix(contractAddress)
 	vmStore := types.NewStoreAdapter(prefix.NewStore(runtime.KVStoreAdapter(k.storeService.OpenKVStore(sdkCtx)), prefixStoreKey))
 	gasLeft := k.runtimeGasForContract(sdkCtx)
-	res, gasUsed, err := k.wasmVM.Migrate(newChecksum, env, msg, vmStore, cosmwasmAPI, &querier, k.gasMeter(sdkCtx), gasLeft, costJSONDeserialization)
+
+	migrateInfo := wasmvmtypes.MigrateInfo{
+		Sender:          senderAddress.String(),
+		OldStateVersion: OldStateVersion,
+	}
+	var res *wasmvmtypes.ContractResult
+	var gasUsed uint64
+	var err error
+	if OldStateVersion == nil {
+		res, gasUsed, err = k.wasmVM.Migrate(newChecksum, env, msg, vmStore, cosmwasmAPI, &querier, k.gasMeter(sdkCtx), gasLeft, costJSONDeserialization)
```

Comment:
I would also say we should add a comment explaining why it is okay to do the second call if the first one errors. When you don't know anything about how `Migrate` and `Migrate2` work under the hood you might mistakenly think that this would migrate twice in some circumstances.
Something like:
> Migrate and Migrate2 call the same contract entrypoint, but with different parameters, so one of them will always fail without actually calling into the contract. Because of that, it is okay to try the other one in case of an error.

Reply:
This is already solved by using only the new MigrateWithInfo call